### PR TITLE
Kad

### PR DESCRIPTION
**Pull request by art-gor**

### CODE REVIEW

File: example/02-kad/factory.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#ifndef LIBP2P_KAD_EXAMPLE_FACTORY_HPP
+#define LIBP2P_KAD_EXAMPLE_FACTORY_HPP
+
+#include <libp2p/crypto/crypto_provider.hpp>
+#include <libp2p/crypto/key_marshaller.hpp>
+#include <libp2p/host/host.hpp>
+#include <memory>
+
+namespace libp2p::kad_example {
+  std::shared_ptr<boost::asio::io_context> createIOContext();
```
Comment (kamilsa) [inline example/02-kad/factory.hpp]: io_context is created by injector. Why do we need another one? Example: https://github.com/soramitsu/libp2p/blob/master/example/01-echo/libp2p_echo_client.cpp#L37

File: include/libp2p/network/impl/dialer_impl.hpp [hunk 2]
```diff
@@ -33,6 +33,8 @@ namespace libp2p::network {
     std::shared_ptr<protocol_muxer::ProtocolMuxer> multiselect_;
     std::shared_ptr<TransportManager> tmgr_;
     std::shared_ptr<ConnectionManager> cmgr_;
+
+    common::Logger log_ = common::createLogger("debug");  // XXX
```
Comment (kamilsa) [inline include/libp2p/network/impl/dialer_impl.hpp]: what does it mean XXX?

File: include/libp2p/protocol/kademlia/impl/asio_scheduler_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  class AsioSchedulerImpl : public Scheduler {
+   public:
+    AsioSchedulerImpl(boost::asio::io_context &io, Ticks interval)
+        : Scheduler{},
+          io_(io),
+          timer_(io, boost::posix_time::milliseconds(interval)),
+          interval_(interval),
+          started_(boost::posix_time::microsec_clock::local_time()),
+          timer_cb_([this](const boost::system::error_code &) { onTimer(); }),
+          immediate_cb_([this] { onImmediate(); }) {
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/asio_scheduler_impl.hpp]: Do you really need those two callbacks?  If you want to invoke onTimer or onImmediate from timer you can use boost::bind for that. Example: https://github.com/soramitsu/kagome/pull/236/files#diff-e4975f9d1f0019308c8ea6f52a3c0e5eR115

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 4]
```diff
@@ -6,36 +6,101 @@
...
+    };
+
+    struct Session {
+      KadProtocolSession::Ptr protocol_handler;
+
+      // nullptr for server sessions
+      KadResponseHandler::Ptr response_handler;
+    };
+
+    typedef bool (KadImpl::*RequestHandler)(Message &msg);
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: What is that? Sorry, not clear for me why it is here

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 5]
```diff
@@ -6,36 +6,101 @@
...
 
 namespace libp2p::protocol::kademlia {
 
-  class KadImpl : public Kad {
+  class KadImpl : public Kad,
+                  public KadSessionHost,
+                  public std::enable_shared_from_this<KadImpl> {
    public:
-    ~KadImpl() override = default;
+    KadImpl(Host &host, std::shared_ptr<RoutingTable> table,
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: Can we use shared_ptr<Host> instead reference? It is more consistent with the rest of the project and more thread safe IMO
```suggestion
    KadImpl(std::shared_ptr<Host> host, std::shared_ptr<RoutingTable> table,
```

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 6]
```diff
@@ -6,36 +6,101 @@
...
+    struct Session {
+      KadProtocolSession::Ptr protocol_handler;
+
+      // nullptr for server sessions
+      KadResponseHandler::Ptr response_handler;
+    };
+
+    typedef bool (KadImpl::*RequestHandler)(Message &msg);
+
+    Session *findSession(connection::Stream *from);
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: can we avoid using raw pointers and prefer smart pointers instead?

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 7]
```diff
@@ -6,36 +6,101 @@
...
+
+    void onMessage(connection::Stream *from, Message &&msg) override;
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: same

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 8]
```diff
@@ -6,36 +6,101 @@
...
+
+    void onCompleted(connection::Stream *from,
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: same

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp [hunk 9]
```diff
@@ -6,36 +6,101 @@
...
+                     outcome::result<void> res) override;
+
+    const KademliaConfig &config() override;
+
+    void newServerSession(std::shared_ptr<connection::Stream> stream);
+
+    void closeSession(connection::Stream *s);
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: same

File: include/libp2p/protocol/kademlia/kad.hpp [hunk 10]
```diff
@@ -7,27 +7,57 @@
 #define LIBP2P_KADEMLIA_KAD_HPP
 
 #include <libp2p/network/network.hpp>
+#include <libp2p/protocol/base_protocol.hpp>
+#include <libp2p/protocol/kademlia/common.hpp>
 #include <libp2p/protocol/kademlia/config.hpp>
-#include <libp2p/protocol/kademlia/value_store.hpp>
-#include <libp2p/routing/content_routing.hpp>
-#include <libp2p/routing/peer_routing.hpp>
+//#include <libp2p/protocol/kademlia/value_store.hpp>
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/kad.hpp]: remove comments

File: include/libp2p/protocol/kademlia/kad.hpp [hunk 11]
```diff
@@ -7,27 +7,57 @@
...
+  };
+
   /**
    * @class Kad
    *
    * Entrypoint to a kademlia network.
    */
-  struct Kad
-      : public PeerRouting /*, public ContentRouting, public ValueStore */ {
+  class Kad : public protocol::BaseProtocol {
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/kad.hpp]: Please leave todos with tasks about ValueStore and ContentRouting

File: include/libp2p/protocol/kademlia/node_id.hpp [hunk 12]
```diff
@@ -53,6 +51,10 @@ namespace libp2p::protocol::kademlia {
    public:
     explicit NodeId(const Hash256 &h) : data_(h) {}
 
+    explicit NodeId(const void *bytes) {
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/node_id.hpp]: please [ByteArray](https://sourcegraph.com/github.com/soramitsu/libp2p/-/blob/include/libp2p/common/byteutil.hpp#L13) type instead

File: include/libp2p/protocol/kademlia/node_id.hpp [hunk 13]
```diff
@@ -53,6 +51,10 @@ namespace libp2p::protocol::kademlia {
...
+      memcpy(data_.data(), bytes, 32);
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/node_id.hpp]: ODR. Shouldn't be moved to CPP?

File: include/libp2p/protocol/kademlia/scheduler.hpp [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    // Lifetime-aware scheduler handle
+    class Handle {
+     public:
+      ~Handle() {
+        cancel();
+      }
+
+      void cancel() {
+        if (scheduler_) {
+          scheduler_->cancel(ticket_);
```
Comment (kamilsa) [inline include/libp2p/protocol/kademlia/scheduler.hpp]: ODR. PLease move to CPP file

File: src/muxer/yamux/yamuxed_connection.cpp [hunk 15]
```diff
@@ -462,6 +467,15 @@ namespace libp2p::connection {
     if (auto stream = findStream(stream_id)) {
       streams_.erase(stream_id);
       stream->resetStream();
+
+      // TODO temporarily cleanup itself!
```
Comment (kamilsa) [inline src/muxer/yamux/yamuxed_connection.cpp]: please leave task with TODO and explain why we need this comments

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+    std::optional<Message::Peer> assign_peer(const kad::pb::Message_Peer &src) {
+      using R = std::optional<Message::Peer>;
+
+      if (int(src.connection()) > int(ConnStatus::CAN_NOT_CONNECT)) {
+        // TODO log
+        return R();
+      }
+
+      auto peer_id_res = libp2p::peer::PeerId::fromBase58(src.id());
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]:
```suggestion
      auto peer_id_res = peer::PeerId::fromBase58(src.id());
```

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+      if (!peer_id_res) {
+        // TODO log
+        return R();
+      }
+
+      std::vector<multi::Multiaddress> addresses;
+      for (const auto &addr : src.addrs()) {
+        auto res = libp2p::multi::Multiaddress::create(addr);
+        if (!res) {
+          // TODO log
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]: fix todo or leave the task

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+    bool assign_peers(std::optional<Message::Peers> &dst,
+                      const PbContainer &src) {
+      if (!src.empty()) {
+        dst = Message::Peers{};
+        Message::Peers &v = dst.value();
+        v.reserve(src.size());
+        for (const auto &p : src) {
+          auto res = assign_peer(p);
+          if (!res) {
+            // TODO log
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]: same

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    size_t msg_sz = pb_msg.ByteSizeLong();
+    auto varint_len = multi::UVarint{msg_sz};
+    auto varint_vec = varint_len.toVector();
+    size_t prefix_sz = varint_vec.size();
+    buffer.resize(prefix_sz + msg_sz);
+    memcpy(buffer.data(), varint_vec.data(), prefix_sz);
+    return pb_msg.SerializeToArray(buffer.data() + prefix_sz, msg_sz);
+  }
+
+  void Message::selfAnnounce(libp2p::peer::PeerInfo self) {
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]:
```suggestion
  void Message::selfAnnounce(peer::PeerInfo self) {
```

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+      assign_blob(dst.value, src.value());
+      dst.time_received = src.timereceived();
+    }
+    if (!assign_peers(closer_peers, pb_msg.closerpeers())) {
+      return false;
+    }
+    return assign_peers(provider_peers, pb_msg.providerpeers());
+  }
+
+  bool Message::serialize(std::vector<uint8_t> &buffer) const {
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]: pls use ByteArray type

File: src/protocol/kademlia/impl/kad_message.cpp [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+        auto res = libp2p::multi::Multiaddress::create(addr);
+        if (!res) {
+          // TODO log
+          return R();
+        }
+        addresses.push_back(res.value());
+      }
+
+      return Message::Peer{
+          libp2p::peer::PeerInfo{peer_id_res.value(), std::move(addresses)},
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_message.cpp]:
```suggestion
          peer::PeerInfo{peer_id_res.value(), std::move(addresses)},
```

File: src/protocol/kademlia/impl/kad_protocol_session.cpp [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+      std::weak_ptr<KadSessionHost> host,
+      std::shared_ptr<connection::Stream> stream)
+      : host_(std::move(host)), stream_(std::move(stream)) {}
+
+  bool KadProtocolSession::read() {
+    if (reading_ || stream_->isClosedForRead() || host_.expired()) {
+      return false;
+    }
+    reading_ = true;
+    libp2p::basic::VarintReader::readVarint(
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/kad_protocol_session.cpp]:
```suggestion
    basic::VarintReader::readVarint(
```

File: src/protocol/kademlia/impl/scheduler.cpp [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  Scheduler::Handle Scheduler::schedule(Scheduler::Ticks delay,
+                                        Scheduler::Callback cb) {
+    assert(cb);
+    auto abs_delay = (delay != 0) ? now() + delay : 0;
+    Ticket ticket(abs_delay, ++counter_);
+    Handle h(new scheduler::Handle(ticket, this, std::move(cb)));
+    table_[ticket] = h.get();
+    //    log_->info("Timer create: now={}, table size={}", last_tick_,
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/scheduler.cpp]: why commented

Comment (kamilsa) [reply src/protocol/kademlia/impl/scheduler.cpp]: same for the rest of the file

File: src/protocol/kademlia/impl/scheduler.cpp [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  void Scheduler::pulse(bool immediate) {
+    if (!immediate) {
+      last_tick_ = now();
+    }
+    for (;;) {
+      //   log_->info("Timer pulse: now={}, table size={}", last_tick_,
+      //   table_.size());
+      scheduler::Handle *h = getHandle(immediate);
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/scheduler.cpp]: prefer smart pointers

File: src/protocol/kademlia/impl/scheduler.cpp [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    auto it = table_.begin();
+    auto time = it->first.first;
+    if (immediate) {
+      if (time != 0) {
+        return nullptr;
+      }
+    } else if (time > last_tick_) {
+      return nullptr;
+    }
+    scheduler::Handle *h = it->second;
```
Comment (kamilsa) [inline src/protocol/kademlia/impl/scheduler.cpp]: same

File: example/02-kad/factory.hpp (hunk 1 above)
Comment (art-gor) [reply example/02-kad/factory.hpp]: Here is such a thing: io_context is shared among all Host instances in the example, other objects are per-host

File: include/libp2p/protocol/kademlia/impl/asio_scheduler_impl.hpp (hunk 3 above)
Comment (art-gor) [reply include/libp2p/protocol/kademlia/impl/asio_scheduler_impl.hpp]: yes, those callbacks do completely different things

File: include/libp2p/protocol/kademlia/impl/kad_impl.hpp (hunk 8 above)
Comment (art-gor) [reply include/libp2p/protocol/kademlia/impl/kad_impl.hpp]: here, connection::Stream* is a key

**Pull request merged**